module/zfs/zio_checksum.c

```diff
@@ -545,14 +545,34 @@ zio_checksum_error(zio_t *zio, zio_bad_cksum_t *info)
 	uint_t checksum = (bp == NULL ? zio->io_prop.zp_checksum :
 	    (BP_IS_GANG(bp) ? ZIO_CHECKSUM_GANG_HEADER : BP_GET_CHECKSUM(bp)));
 	int error;
-	uint64_t size = (bp == NULL ? zio->io_size :
-	    (BP_IS_GANG(bp) ? SPA_GANGBLOCKSIZE : BP_GET_PSIZE(bp)));
+	uint64_t size = bp == NULL ? zio->io_size :
+	    (BP_IS_GANG(bp) ? SPA_OLD_GANGBLOCKSIZE : BP_GET_PSIZE(bp));
 	uint64_t offset = zio->io_offset;
 	abd_t *data = zio->io_abd;
 	spa_t *spa = zio->io_spa;
 
+	if (bp && BP_IS_GANG(bp) && spa_feature_is_active(spa,
+	    SPA_FEATURE_DYNAMIC_GANG_HEADER)) {
+		size = zio->io_size;
```

Comment:
This nested size logic looks messy, even if correct.

Reply:
The fact that we have it after we already use a ternary to set the size kind of bothers me. The expanded version works, but I'm not sure about how readable it is; thoughts?

```
	if (bp) {
		if (BP_IS_GANG(bp)) {
			if (spa_feature_is_active(spa,
			    SPA_FEATURE_DYNAMIC_GANG_HEADER))
				size = zio->io_size;
			else
				size = SPA_OLD_GANGBLOCKSIZE;
		} else {
			size = BP_GET_PSIZE(bp);
		}
	} else {
		size == zio->io_size;
	}
```